packages/react/src/Image.tsx

```diff
@@ -0,0 +1,403 @@
+import "@uppy/progress-bar/dist/style.css";
+
+import { createReactBlockSpec, InlineContent } from "./ReactBlockSpec";
+
+// Converts text alignment prop values to the flexbox `align-items` values.
+const textAlignmentToAlignItems = (
+  textAlignment: "left" | "center" | "right" | "justify"
+): "flex-start" | "center" | "flex-end" => {
+  switch (textAlignment) {
+    case "left":
+      return "flex-start";
+    case "center":
+      return "center";
+    case "right":
+      return "flex-end";
+    default:
+      return "flex-start";
+  }
+};
+
+// Max & min image widths as a fraction of the editor's width
+const maxWidth = 1.0;
+const minWidth = 0.1;
+
+const imagePropSchema = {
+  textAlignment: defaultProps.textAlignment,
+  backgroundColor: defaultProps.backgroundColor,
+  // Image src
+  src: {
+    // TODO: Better default
+    default: "https://via.placeholder.com/150" as const,
+  },
+  // Image width as a fraction of the editor's width
+  width: {
+    default: "0.5" as const,
+  },
+  // Whether to show the image upload dashboard or not
+  replacing: {
+    default: "false" as const,
+    values: ["true", "false"] as const,
+  },
+} satisfies PropSchema;
+
+const ImageComponent = <Caption extends boolean>(props: {
+  block: Caption extends true
+    ? SpecificBlock<
+        BlockSchema & {
+          captionedImage: BlockSpec<
+            "captionedImage",
+            typeof imagePropSchema,
+            Caption
+          >;
+        },
+        "captionedImage"
+      >
+    : SpecificBlock<
+        BlockSchema & {
+          image: BlockSpec<"image", typeof imagePropSchema, Caption>;
+        },
+        "image"
+      >;
+  editor: Caption extends true
+    ? BlockNoteEditor<
+        BlockSchema & {
+          captionedImage: BlockSpec<
+            "captionedImage",
+            typeof imagePropSchema,
+            Caption
+          >;
+        }
+      >
+    : BlockNoteEditor<
+        BlockSchema & {
+          image: BlockSpec<"image", typeof imagePropSchema, Caption>;
+        }
+      >;
+  caption: Caption;
+}) => {
+  // Used to check if the resizing handles should be shown.
+  const [hovered, setHovered] = useState(false);
+
+  // Used to check if the image is being resized, and which resize handle is
+  // being used (left or right).
+  const [resizeHandle, setResizeHandle] = useState<"left" | "right" | null>(
+    null
+  );
+
+  // Used to calculate the new width while resizing the image, as the cursor X
+  // offset is just added to the initial width. Both values are represented in
+  // px.
+  const [initialWidth, setInitialWidth] = useState(0);
+  const [initialClientX, setInitialClientX] = useState(0);
+
+  // The editor's width in px.
+  const [editorWidth, setEditorWidth] = useState(
+    props.editor.domElement.firstElementChild!.clientWidth
+  );
+
+  // The image width, represented as a fraction of the editor's width.
+  const [width, setWidth] = useState(() => parseFloat(props.block.props.width));
```

Comment:
Interesting, we need to think about this datamodel. How do we want to store the image width?

If you store it as a fraction of the editor's width, this kind of implies that documents will always be shown at the same size. What if the user resizes the image to 50%, but then opens the document on mobile? Then it will be quite small right? Wondering how editors like notion handle this

Reply:
I went with this approach since I figured it would work better for different screen sizes than e.g. storing the width in px, since otherwise images could go beyond the editor bounds or be really too small. Either way, imo not a huge deal rn since mobile UX is a whole rabbit hole that I think we'll look into some time in the future.

Reply:
Also not a big fan of how Notion does this - it broadly works the same way, but also crops the image if you try to increase the size beyond the editor and their cropping UI is very choppy. Also it breaks sometimes when snapping a window